### PR TITLE
Monorepo Utils: Add workflow profiler command

### PR DESCRIPTION
**Pull request by psealock**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds a CLI command to profile workflows.

image

#### Assumptions made

* Only successful workflow run are used to compile results
* All workflow runs lasting longer than 60 minutes are excluded.
* 90th percentile means the 90% slowest run

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. `pnpm install` to make sure the tool is correctly built.
2. Try running the tool for `Lint and tests for JS packages and woocommerce-admin/client`.
```
pnpm utils workflows profile 2023-06-01 2023-06-21 pr-lint-test-js.yml
```
2. Try again but pass `-s` to show job and step results too.
```
pnpm utils workflows profile 2023-06-01 2023-06-21 pr-lint-test-js.yml -s
```
3. Now run the tool for `Run tests against PR`, a workflow that has 6 jobs.
```
pnpm utils workflows profile 2023-06-20 2023-06-20 pr-build-and-e2e-tests.yml 
```
4. Now try running the tool with a fair amount of worflows. Here is `Lint and tests for JS packages and woocommerce-admin/client` for the month of May with 1003 runs.
```
pnpm utils workflows profile 2023-05-01 2023-05-31 pr-lint-test-js.yml -s
```

### CODE REVIEW

File: tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+} from '../../lib/log';
+import { Logger } from '../../../core/logger';
+
+const program = new Command( 'profile' )
+	.description( 'Profile GitHub workflows' )
+	.argument( '<start>', 'Start date in YYYY-MM-DD format' )
+	.argument( '<end>', 'End date in YYYY-MM-DD format' )
+	.argument(
+		'<id>',
+		'Workflow ID. The required workflow ids are 22745783, 5687250, 23271226, and 5461563. For the rest, use the `list` command to see workflow names and ids.'
```
Comment (ObliviousHarmony) [inline tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts]: These IDs are going to be prone to drifting since they likely won't be updated here if we change the workflows. Also, I wonder, do these IDs ever change?

File: tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	logStepResults,
+} from '../../lib/log';
+import { Logger } from '../../../core/logger';
+
+const program = new Command( 'profile' )
+	.description( 'Profile GitHub workflows' )
+	.argument( '<start>', 'Start date in YYYY-MM-DD format' )
+	.argument( '<end>', 'End date in YYYY-MM-DD format' )
+	.argument(
+		'<id>',
```
Comment (ObliviousHarmony) [inline tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts]: Having to use `list` and then `profile` with an ID feels unnecessarily granular. How would you feel about using the workflow file instead and pulling the ID from that ourselves? Is that possible?

Comment (ObliviousHarmony) [reply tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts]: This approach would also make the `list` command unnecessary. Personally, I find the output of that a little hard to follow since it's hard to understand what workflows _really_ do what. The specific events that a workflow runs on give more context about the workflow itself.

File: tools/monorepo-utils/src/workflow-profiler/lib/github.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	const { data } = await octokitWithAuth().request( endpoint, {
+		...requestOptions,
+		page,
+		per_page,
+	} );
+
+	let resultingTotals = processPage( data, totals );
+
+	const { count_items_available } = data;
+	if ( count_items_available > resultingTotals.count_items_processed ) {
```
Comment (ObliviousHarmony) [inline tools/monorepo-utils/src/workflow-profiler/lib/github.ts]:
```suggestion
	if ( resultingTotals.total_count > resultingTotals.count_items_processed ) {
```

The `data` object doesn't actually have a `total_count` property. This is causing pagination to break.

**Review by ObliviousHarmony — CHANGES_REQUESTED** [review]
Thanks @psealock, I really like the table output here, I think it gives a ton of valuable information. I've identified some problems in my testing and left feedback about that. Looking at the tool holistically though, I have some thoughts. How would you feel about changing this a little so that you get the least-precise output that you're looking for? Maybe something like:

- `pnpm utils workflows profile 2023-05-01 2023-05-31`: Returns workflow run results for all workflows (identify slowest workflows)
- `pnpm utils workflows profile 2023-05-01 2023-05-31 workflow-file.yml`: Returns job results for workflow (identify slowest jobs)
- `pnpm utils workflows profile 2023-05-01 2023-05-31 workflow-file.yml job-name-or-such`: Returns step results for job (identify slowest steps)

I came away from my testing feeling like we are given a _lot_ of data that might not be needed in all cases. The above might be excessive though. At the very least, maybe we should put the steps behind a `--show-steps` option or something?

File: tools/monorepo-utils/src/workflow-profiler/lib/github.ts (hunk 3 above)
Comment (ObliviousHarmony) [reply tools/monorepo-utils/src/workflow-profiler/lib/github.ts]: As an aside, this issue kind of stopped my review because when I fixed it I started getting API errors that likely need looking at and that may change the code a bit 😄

File: tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts (hunk 2 above)
Comment (psealock) [reply tools/monorepo-utils/src/workflow-profiler/commands/profile/index.ts]: > The specific events that a workflow runs on give more context about the workflow itself.

Cool idea. I'd need to parse the file itself to get that info.

**Comment by psealock** [comment]
Thanks for the review @ObliviousHarmony 

Here are the changes I made. The testing instructions have been updated.
* The GraphQL request is now paginated
* Fixed count/pagination refactor error
* You can now pass in an id or workflow filename
* There is a `--show-steps` option to fetch and show jobs and step results
* The `list` command now shows filenames as well
* Better logging so you can see progress

I did not implement the suggestion to be able to pass in a job name to zero in on that. Let me know how this goes and happy to add that in if need be.

**Comment by ObliviousHarmony** [comment]
Thanks @psealock, I've taken another look at this PR. It looks like you may have forgotten to push your latest changes? I'm getting an error that the `-s` command doesn't exist, and in the help text, I'm not seeing it either.

**Review by ObliviousHarmony — APPROVED** [review]
Very neat @psealock, thanks for taking care of this. Everything works as you've described.
